### ISSUE TITLE
docker wait '--state' option

### ISSUE THREAD

**Issue by flungo** [docker/docker#18993]
Feature request for the addition of a `--state` (`-s`) option in the `docker wait` command allowing any valid state to be waited on including:

- `created`
- `running`
- `paused`
- `stopped` (**default** in order to maintain backwards compatibility) 
- `deleted`

This would be really useful in scripting environments - the current method that I am using is the following one liner (based off a suggestion from #7228):

```
until [ "$(docker inspect -f {{.State.Running}} ${DOCKER_CONTAINER})" == "true" ]; do sleep 0.1; done
```

Which under my suggestion, would be equivalent to:

```
docker wait --state running ${DOCKER_CONTAINER}
```

This should be fairly easy to implement given the existing events system. One thing that would be required would be the ability for the container specified to `docker wait` to not exist yet - possibly restricted/defaulted based on which state is chosen or with another flag that would whether to ignore if the container doesn't exist yet: the `created` state, for example, could *never* be waited upon if the container had to exist.

# References

- https://docs.docker.com/engine/reference/commandline/wait/
- https://github.com/docker/docker/issues/7228#issuecomment-50108257

**Comment by thaJeztah** [comment]
Thanks for proposing

/cc @calavera, who's working on a rewrite of the events system

**Comment by wmark** [comment]
To further illustrate the need in a use-case, which modifies network settings beyond what Docker can achieve:

```
[Service]
ExecStart=docker run … --name my-voip

ExecStartPost=-/bin/bash -c "iptables -t mangle -I PREROUTING -p all -s $(docker inspect --format '{{ .NetworkSettings.IPAddress }}' my-voip) -j DSCP --set-dscp-class EF"
```

Obviously there's without ```docker wait --state X``` a race condition between ```docker run``` and ```docker inspect``` and ExecStartPost sometimes fails.

TL;DR:

 * add option ```--timeout``` to ```docker wait --state X```
 * don't fail if the container does not exist yet

**Comment by flungo** [comment]
That is pretty much the exact scenario I would be using this in, the workaround I am using is putting the following before the `iptables` command:

```
ExecStartPost=/usr/bin/bash -c 'until [ "$(docker inspect -f {{.State.Running}} ${DOCKER_CONTAINER})" == "true" ]; do sleep 0.1; done'
```

You can find more examples of uses with the drop-ins that are in [this project](https://github.com/SystemDocker/common/tree/master/dropins).

Timeout and working on non-existent containers are also both crucial for this to be useful.

**Comment by thaJeztah** [comment]
We also have health checks no, which could be helpful for your current approach.

We discussed this, and think your current approach is working, so don't consider this a high priority. I'm closing this issue, but feel free to comment after
